### ISSUE TITLE
Feature request: Improve import assist by considering types

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-analyzer#10471]
Today I (re)wrote some code like this:

```rust
match channel_sender.try_send(value) {
    Err(TrySendError::Full(_)) => todo!(),
    // ...
}
```

And used <kbd>Ctrl</kbd>+<kbd>.</kbd> to import `TrySendError`. There were multiple choices due to dependencies on all of `std`, `tokio` and `async-channel`. It would have been nice to not have to navigate the `Import TrySendError` sub-menu where I had to choose between the various available types, since RA correctly figured out that `try_send` is from tokio and returns `Result<(), tokio::sync::mpsc::error::TrySendError<T>>`.

**Comment by jplatte** [comment]
Hey @jmintb, I saw you did #15627 (awesome work!) which I initially thought would work for the example in this issue, based on the title of the PR. I guess it just works when the expected type comes from a function argument type, which is what the issue that your PR closed was about?

Since it seems like a very closely related issue, do you have any plans for working on this too? Just asking to avoid duplicate work since I might otherwise take a stab at it soon :)

**Comment by Veykril** [comment]
That PR was about completions, not about the assist which currently ignores expected types in general.

**Comment by jplatte** [comment]
Oh, I see! That kind of sounds like a refactoring is in order, since these two features should generally show the same lists of symbols, no? (except maybe flyimport should limit how many different ones are offered)

**Comment by Veykril** [comment]
Yes, the completions stuff already has the logic, we'd "just" need to move that out (to `ide-db`) and have both make use of that.

**Comment by jmintb** [comment]
> Hey @jmintb, I saw you did #15627 (awesome work!) which I initially thought would work for the example in this issue, based on the title of the PR. I guess it just works when the expected type comes from a function argument type, which is what the issue that your PR closed was about?
> 
> Since it seems like a very closely related issue, do you have any plans for working on this too? Just asking to avoid duplicate work since I might otherwise take a stab at it soon :)

I wasn't aware of this issue, but I am up for working on this next. :)

**Comment by jplatte** [comment]
Awesome! I haven't started anything yet.
